File: Back/FastAPI/routers/screener.py
"""
routers/screener.py
===================
다중 조건 종목 스크리너 엔드포인트.

ML 점수 + 재무 지표 조건을 조합하여 종목을 필터링합니다.
"""
from __future__ import annotations

from typing import Optional
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from ..services import data as svc

router = APIRouter(prefix="/screener", tags=["screener"])


class ScreenerItem(BaseModel):
    ticker:         str
    name:           Optional[str]   = None
    sector:         Optional[str]   = None
    close:          Optional[float] = None   # 현재가 (scores.close)
    change_pct:     Optional[float] = None   # 전일 대비 등락률
    score:          Optional[float] = None
    tier:           Optional[str]   = None
    signal_label:   Optional[str]   = None   # 매수/보유/매도/관망
    latest_date:    Optional[str]   = None
    per:            Optional[float] = None
    pbr:            Optional[float] = None
    roe:            Optional[float] = None
    debt_ratio:     Optional[float] = None
    op_margin:      Optional[float] = None
    rev_growth_yoy: Optional[float] = None
    finance_score:  Optional[float] = None
    composite_score: Optional[float] = None  # ML 60% + Finance 40%


class ScreenerResponse(BaseModel):
    total:  int
    items:  list[ScreenerItem]


@router.get("", response_model=ScreenerResponse, summary="다중 조건 종목 스크리너")
def screen_stocks(
    # ML 조건
    model_version:  str            = Query("latest", description="모델 버전"),
    min_score:      float          = Query(0.0, ge=0, le=100, description="최소 ML 점수"),
    # 422 가드 — 부정값은 silent fallback 대신 명시적으로 거절.
    tier:           Optional[str]  = Query(None, pattern="^[ABCD]$", description="티어 필터: A | B | C | D"),
    sector:         Optional[str]  = Query(None, max_length=80, description="섹터 이름 (부분 일치, 최대 80자)"),
    # 재무 조건
    max_per:        Optional[float] = Query(None, description="최대 PER"),
    max_pbr:        Optional[float] = Query(None, description="최대 PBR"),
    min_roe:        Optional[float] = Query(None, description="최소 ROE (%)"),
    max_debt_ratio: Optional[float] = Query(None, description="최대 부채비율 (%)"),
    min_op_margin:  Optional[float] = Query(None, description="최소 영업이익률 (%)"),
    min_rev_growth: Optional[float] = Query(None, description="최소 매출성장률 YoY (%)"),
    min_finance_score: Optional[float] = Query(None, ge=0, le=100, description="최소 재무 스코어"),
    # 정렬·페이지 — 알려진 컬럼만 허용 (이전엔 부정값이 silent 하게 composite_score 로 fallback)
    sort_by:        str            = Query(
        "composite_score",
        pattern="^(composite_score|score|finance_score|roe|per|pbr|rev_growth_yoy)$",
        description="정렬 기준: composite_score | score | finance_score | roe | per | pbr | rev_growth_yoy",
    ),
    ml_weight:      float          = Query(
        0.6, ge=0.0, le=1.0,
        description="B11: composite_score 의 ML 가중치 (기본 0.6, finance 가중치 = 1 - ml_weight).",
    ),
    limit:          int            = Query(50, ge=1, le=200, description="최대 결과 수"),
):
    """
    ML 추천 점수 + 재무 지표를 복합 조건으로 종목을 검색합니다.

    **composite_score** = ML score × 0.6 + finance_score × 0.4
    """
    try:
        rows = svc.screen_stocks(
            model_version=model_version,
            min_score=min_score,
            tier=tier,
            sector=sector,
            max_per=max_per,
            max_pbr=max_pbr,
            min_roe=min_roe,
            max_debt_ratio=max_debt_ratio,
            min_op_margin=min_op_margin,
            min_rev_growth=min_rev_growth,
            min_finance_score=min_finance_score,
            sort_by=sort_by,
            limit=limit,
            ml_weight=ml_weight,
        )
    except RuntimeError as e:
        raise HTTPException(status_code=503, detail=str(e))

    items = [ScreenerItem(**r) for r in rows]
    return ScreenerResponse(total=len(items), items=items)

File: Front/src/stores/screener.js
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { stocksApi } from '@/api/stocks.js'

const TIER_MIN_SCORE = { A: 80, B: 60, C: 40, D: 0 }

export const useScreenerStore = defineStore('screener', () => {
  const items   = ref([])
  const total   = ref(0)
  const loading = ref(false)
  const error   = ref(null)

  const filters = ref({
    model_version:     'latest',
    min_score:         0,
    tier:              null,
    sector:            null,
    max_per:           null,
    max_pbr:           null,
    min_roe:           null,
    max_debt_ratio:    null,
    min_op_margin:     null,
    min_rev_growth:    null,
    min_finance_score: null,
    sort_by:           'composite_score',
    limit:             50,
  })

  async function fetchScreener() {
    loading.value = true
    error.value   = null
    try {
      const f = filters.value
      const params = {}
      if (f.model_version) params.model_version = f.model_version
      if (f.sector)        params.sector        = f.sector
      if (f.tier)          params.tier          = f.tier
      // tier → min_score 변환
      const tierMin  = f.tier ? (TIER_MIN_SCORE[f.tier] ?? 0) : 0
      const minScore = Math.max(Number(f.min_score ?? 0), tierMin)
      if (minScore > 0) params.min_score = minScore
      if (f.max_per)           params.max_per           = f.max_per
      if (f.max_pbr)           params.max_pbr           = f.max_pbr
      if (f.min_roe)           params.min_roe           = f.min_roe
      if (f.max_debt_ratio)    params.max_debt_ratio    = f.max_debt_ratio
      if (f.min_op_margin)     params.min_op_margin     = f.min_op_margin
      if (f.min_rev_growth)    params.min_rev_growth    = f.min_rev_growth
      if (f.min_finance_score) params.min_finance_score = f.min_finance_score
      params.sort_by = f.sort_by ?? 'composite_score'
      params.limit   = f.limit   ?? 50
      const { data } = await stocksApi.getScreener(params)
      items.value = data.items ?? []
      total.value = data.total ?? 0
    } catch (e) {
      error.value = 'error'
      items.value = []
    } finally {
      loading.value = false
    }
  }

  function setFilter(key, value) {
    filters.value[key] = value === '' ? null : value
  }

  function resetAllFilters() {
    filters.value.max_per           = null
    filters.value.max_pbr           = null
    filters.value.min_roe           = null
    filters.value.max_debt_ratio    = null
    filters.value.min_op_margin     = null
    filters.value.min_rev_growth    = null
    filters.value.min_finance_score = null
    filters.value.min_score         = 0
    filters.value.tier              = null
    filters.value.sector            = null
  }

  return { items, total, loading, error, filters, fetchScreener, setFilter, resetAllFilters }
})
